compiler/src/org.graalvm.compiler.asm.aarch64/src/org/graalvm/compiler/asm/aarch64/AArch64Assembler.java

```diff
@@ -3330,44 +3277,2087 @@ public void patch(int codePos, int relative, byte[] code) {
+         * @param eSize size of value to duplicate.
+         * @param dst SIMD register.
+         * @param src SIMD register.
+         * @param index offset of value to duplicate
+         */
+        public void dupVector(ASIMDSize dstSize, ElementSize eSize, Register dst, Register src, int index) {
+            assert src.getRegisterCategory() == SIMD;
+            assert dst.getRegisterCategory() == SIMD;
+
+            asimdCopy(ASIMDInstruction.DUPELEM, dstSize == ASIMDSize.FullReg, eSize, dst, src, index);
+        }
+
+        /**
+         * C7.2.371 Unsigned move vector element to general-purpose register.<br>
+         *
+         * <code>dst (gp) = src[index] (simd).</code>
+         *
+         * @param eSize width of element to move.
+         * @param dst general-purpose register.
+         * @param src SIMD register.
+         * @param index offset of value to move.
+         */
+        public void umovGeneral(ElementSize eSize, Register dst, Register src, int index) {
+            assert dst.getRegisterCategory() == CPU;
+            assert src.getRegisterCategory() == SIMD;
+
+            asimdCopy(ASIMDInstruction.UMOV, eSize == ElementSize.DoubleWord, eSize, dst, src, index);
+        }
+
+        /**
+         * C7.2.279 Signed move vector element to general-purpose register.<br>
+         *
+         * <code>dst (gp) = sign-extend(src[index]) (simd).</code>
+         *
+         * Note that the target register size (dst) must be greater than the source element size.
+         *
+         * @param dstESize width of sign-extended.
+         * @param srcESize width of element to move.
+         * @param dst general-purpose register.
+         * @param src SIMD register.
+         * @param index offset of value to move.
+         */
+        public void smovGeneral(ElementSize dstESize, ElementSize srcESize, Register dst, Register src, int index) {
+            assert srcESize != ElementSize.DoubleWord;
+            assert dstESize == ElementSize.Word || dstESize == ElementSize.DoubleWord;
+            assert srcESize.nbits < dstESize.nbits : "the target size must be larger than the source size";
+
+            assert dst.getRegisterCategory() == CPU;
+            assert src.getRegisterCategory() == SIMD;
+
+            asimdCopy(ASIMDInstruction.SMOV, dstESize == ElementSize.DoubleWord, srcESize, dst, src, index);
+        }
+
+        /**
+         * C7.2.2 Integer add scalar.<br>
+         *
+         * <code>dst[0] = int_add(src1[0], src2[0])</code>
+         *
+         * Note that only 64-bit (DoubleWord) operations are available.
+         *
+         * @param eSize element size. Must be of type ElementSize.DoubleWord
+         * @param dst SIMD register.
+         * @param src1 SIMD register.
+         * @param src2 SIMD register.
+         */
+        public void addScalar(ElementSize eSize, Register dst, Register src1, Register src2) {
+            assert dst.getRegisterCategory() == SIMD;
```

Comment:
Yeah, I also fear the potential bugs lurking here. I still need to very very carefully look over everything again. So far, I've tried to be pretty strict in structure to try to minimize the potential for bugs, and initial testing _seems_ okay, but there should be something more thorough.

I like your annotation idea. I think something like
@ASIMD(op="add", category=THREE_SAME) which then automatically generates a couple instructions, feeds them to a disassembler, and checks the name with the expected format would be pretty cool.

Reply:
Sounds great, as long as THREE_SAME means three register of the same type, and not the exact same register three times. So that you can check add r1,r2,r3 doesn't become add r2,r3,r1